### PR TITLE
Refactor the api key resource

### PR DESCRIPTION
**Pull request by TheMattSchiller**

### CODE REVIEW

File: banyan/resource_api_key_test.go [hunk 1]
```diff
@@ -13,7 +13,7 @@ import (
 // Use the terraform plugin sdk testing framework for example testing apikey lifecycle
 func TestAccApiKey_basic(t *testing.T) {
 
-	rName := acctest.RandStringFromCharSet(10, acctest.CharSetAlphaNum)
+	rName := fmt.Sprintf("tf-acc-%s", acctest.RandStringFromCharSet(10, acctest.CharSetAlphaNum))
```
Comment (danielboros) [inline banyan/resource_api_key_test.go:16]: These tests were running on `dev06`?

File: banyan/resource_api_key.go [hunk 2]
```diff
@@ -39,82 +37,86 @@ func resourceApiKey() *schema.Resource {
...
 				Sensitive:   true,
+				Computed:    true,
 				ForceNew:    true,
 			},
 			"scope": {
-				Type:        schema.TypeString,
-				Description: "API Secret key",
-				Optional:    true,
+				Type:         schema.TypeString,
+				Description:  "Scope for the API key. Must be one of: \"satellite\", \"access_tier\", \"read_logs\", \"Admin\", \"ServiceAuthor\", \"PolicyAuthor\", \"EventWriter\", \"ReadOnly\"",
```
Comment (danielboros) [inline banyan/resource_api_key.go:46]: Is this a more generic API key description or specific to access tier? If it's for access tier, then it needs to be `access_tier` or `Admin` scope.

File: banyan/resource_api_key.go [hunk 3]
```diff
@@ -39,82 +37,86 @@ func resourceApiKey() *schema.Resource {
...
 
 func resourceApiKeyRead(ctx context.Context, d *schema.ResourceData, m interface{}) (diagnostics diag.Diagnostics) {
-	log.Printf("[APIKEY|RES|READ] reading apikey %s : %s", d.Get("name"), d.Id())
-	client := m.(*client.Holder)
-	k, err := client.ApiKey.Get(d.Id())
+	c := m.(*client.Holder)
+	key, err := c.ApiKey.Get(d.Id())
 	if err != nil {
-		return handleNotFoundError(d, fmt.Sprintf("apikey %q", d.Id()))
+		handleNotFoundError(d, err)
```
Comment (danielboros) [inline banyan/resource_api_key.go:90]: Will this always be a not found error?

File: client/apikey/apikey.go [hunk 4]
```diff
@@ -1,123 +1,87 @@
...
-	Update(post Post) (updatedApiKey Data, err error)
+	Update(id string, post Post) (updatedApiKey Data, err error)
 	Delete(id string) (err error)
 }
 
 func (k *ApiKey) Get(id string) (apikey Data, err error) {
-	if id == "" {
-		err = errors.New("need an id to get a apikey")
+	resp, err := k.restClient.Read(apiVersion, component, id, "")
+	if err != nil {
```
Comment (danielboros) [inline client/apikey/apikey.go:34]: Should we also be checking `resp` for error conditions?

Something like:

```go
if resp.StatusCode < 200 || resp.StatusCode >= 300 {
  return fmt.Errorf("got failing status: %d, response: %s", resp.StatusCode, body)
}
```

File: client/apikey/apikey.go [hunk 5]
```diff
@@ -1,123 +1,87 @@
...
-	defer response.Body.Close()
-	if response.StatusCode != 200 {
-		err = errors.Errorf("unsuccessful, got status code %q with response: %+v for request to %s", response.Status, response.Request, path)
-		return
-	}
-	responseData, err := ioutil.ReadAll(response.Body)
-	if err != nil {
+func (k *ApiKey) Create(post Post) (apikey Data, err error) {
+	// check if key exists already
+	responseJSON, err := getAll(k)
```
Comment (danielboros) [inline client/apikey/apikey.go:47]: `err` not checked.

File: client/apikey/apikey.go [hunk 6]
```diff
@@ -1,123 +1,87 @@
...
+	apikey, err = findByName(post.Name, responseJSON)
+	if err == nil {
+		err = errors.Errorf("api key already exists: %s", post.Name)
 		return
 	}
-	err = json.Unmarshal(responseData, &responseJSON)
+	body, err := json.Marshal(post)
```
Comment (danielboros) [inline client/apikey/apikey.go:53]: `err` not checked.

File: client/apikey/apikey.go [hunk 7]
```diff
@@ -1,123 +1,87 @@
...
+	response, err := k.restClient.Create(apiVersion, component, body, "")
 	if err != nil {
 		return
 	}
+	var responseData CreateResponse
+	err = json.Unmarshal(response, &responseData)
```
Comment (danielboros) [inline client/apikey/apikey.go:59]: `err` not checked.

File: client/apikey/apikey.go [hunk 8]
```diff
@@ -1,123 +1,87 @@
...
-		if key.Name == name {
-			apikey = key
-			ok = true
-		}
+func (k *ApiKey) Update(id string, post Post) (updatedApiKey Data, err error) {
+	body, err := json.Marshal(post)
+	if err != nil {
+		return
 	}
+	resp, err := k.restClient.Update(apiVersion, component, id, body, "")
```
Comment (danielboros) [inline client/apikey/apikey.go:69]: `err` not checked.

File: client/apikey/apikey.go [hunk 9]
```diff
@@ -126,53 +90,25 @@ func (k *ApiKey) Create(post Post) (apikey Data, err error) {
...
-	log.Printf("[APIKEY|DELETE] deleting apikey with id %s", apikey.ID)
-	path := fmt.Sprintf("api/experimental/v2/api_key/%s", apikey.ID)
-	myUrl, err := url.Parse(path)
-	if err != nil {
-		return
-	}
-	resp, err := k.restClient.DoDelete(myUrl.String())
-	if err != nil {
-		return
+func findByName(name string, responseJSON Response) (apikey Data, err error) {
```
Comment (danielboros) [inline client/apikey/apikey.go:104]: `Data` is somewhat vague here. Maybe rename model to `APIKeyData` or something more descriptive.

**Review by danielboros — CHANGES_REQUESTED** [review]

File: banyan/resource_api_key_test.go (hunk 1 above)
Comment (TheMattSchiller) [reply banyan/resource_api_key_test.go:16]: yes. I still need to write a sweeper

File: banyan/resource_api_key.go (hunk 2 above)
Comment (TheMattSchiller) [reply banyan/resource_api_key.go:46]: This is a generic resource for the API key so you can make any API key type

File: client/apikey/apikey.go (hunk 4 above)
Comment (TheMattSchiller) [reply client/apikey/apikey.go:34]: ive added this to to restClient.Read

File: client/apikey/apikey.go (hunk 9 above)
Comment (TheMattSchiller) [reply client/apikey/apikey.go:104]: I agree this need to be addressed in the structs for the SDK

**Review by danielboros — APPROVED** [review]

**Pull request merged**